From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Fix MC-157589

The behavior is registered as "present",
causing HIDE behavior never getting timeout if the villager doesn't have a hiding place.

Related MC issue: https://bugs.mojang.com/browse/MC/issues/MC-157589

diff --git a/net/minecraft/world/entity/ai/behavior/SetHiddenState.java b/net/minecraft/world/entity/ai/behavior/SetHiddenState.java
index 5c2b736e33b8bcea33aac6760afdfce1573ff9e8..7194f34915a7ddcc5401d8d84eb290acf91c7a94 100644
--- a/net/minecraft/world/entity/ai/behavior/SetHiddenState.java
+++ b/net/minecraft/world/entity/ai/behavior/SetHiddenState.java
@@ -13,12 +13,16 @@ public class SetHiddenState {
         int stayHiddenTicks = seconds * 20;
         MutableInt ticksHidden = new MutableInt(0);
         return BehaviorBuilder.create(
-            i -> i.group(i.present(MemoryModuleType.HIDING_PLACE), i.present(MemoryModuleType.HEARD_BELL_TIME))
+            i -> i.group(i.registered(MemoryModuleType.HIDING_PLACE), i.present(MemoryModuleType.HEARD_BELL_TIME)) // Leaf - Fix MC-157589
                 .apply(i, (hidingPlace, heardBellTime) -> (level, body, timestamp) -> {
                     long timeTriggered = i.<Long>get(heardBellTime);
                     boolean timedOutTryingToHide = timeTriggered + HIDE_TIMEOUT <= timestamp;
                     if (ticksHidden.intValue() <= stayHiddenTicks && !timedOutTryingToHide) {
-                        BlockPos hidePos = i.get(hidingPlace).pos();
+                        // Leaf start - Fix MC-157589
+                        java.util.Optional<net.minecraft.core.GlobalPos> hidingPlacePos = i.tryGet(hidingPlace);
+                        if (hidingPlacePos.isEmpty()) return false;
+                        // Leaf end - Fix MC-157589
+                        BlockPos hidePos = hidingPlacePos.get().pos();
                         if (hidePos.closerThan(body.blockPosition(), closeEnoughDist)) {
                             ticksHidden.increment();
                         }
